### ISSUE TITLE
Provide a way to use smart rollup chunking for dependencies of manualChunks

### ISSUE THREAD

**Issue by jods4** [rollup/rollup#4180]
### Expected Behavior / Situation

Here's my context:

In a large application, I want to bundle related features into a single chunk.
I have a way to know if a module should be part of a specific chunk and I'm forcing them with a `manualChunk` function in rollup configuration.

Let's say I'm forcing `moduleA1`, `moduleA2` into `chunkA`; `moduleB1` and `moduleB2` into `chunkB`.

Those modules have dependencies, some unique to them (say `privateA`) and some shared (say `shared1`).

What I would like to have, is rollup doing its usual magic dependency graph splitting after I've manually merged some modules.

In my example: `privateA` is used only by `moduleA1` and `moduleA2`, so it can be merged into `chunkA` as well. `shared1` having dependent in both `chunkA` and `chunkB`, it should go into a new `chunkShared` that is imported by both.

This is basically the same thing as rollup would do if I didn't use `manualChunk`.

### Actual Behavior / Situation

As pointed out in the documentation, something slightly different happens.

When `moduleA` is put into `chunkA`, _all its dependencies_ that are not already into another chunk are brought along and put into `chunkA`. This includes `privateA` _and_ `shared1`.

This is simple but not efficient at runtime, and defeats code splitting.

When `moduleB1` is later processed, it sees `shared1` was already put into `chunkA`, so `chunkA` becomes a dependency of `chunkB`.

Consider this situation:
`moduleA1` and `moduleA2` are large; `shared1` is very small.

When the application dynamically imports `moduleB1`; it needs to import the complete, large, `chunkA` just because it needs the small `shared1` module.

In the end, instead of loading just what was needed, everything had to be fetched, including the large, unused `moduleA1`. The benefit of code splitting is lost.

### Modification Proposal

I would like `manualChunk` algorithm to change -- always, or opted into with an option to avoid breaking behavior of existing builds, or even a brand new alternative config.

It should be a first pass that merges some modules forcefully, but _just those modules, no dependencies_.
Then the regular rollup algorithm should run as usual and decide how to best split the whole graph.

**Comment by jods4** [comment]
Small update: the way `manualChunk` works today can produce builds that don't work.
It did today for me, here's what happened:

In my app, I'm using a 3rd party lib (svg icons) that has one side-effect when loaded. This side-effect must run before any function is called otherwise the app crash.

Without chunking, this is no problem. Whenever a module in my app loads the svg lib, it's initialized and then all is fine, as you'd expect.

Partitioning my _app_ (not 3rd parties) modules into, say, chunks A and B breaks this.
Because manual chunking takes with it _any_ dependencies of the first manual chunk that happens to require it, chunk A ended up including the svg lib, and chunks A and B had a circular dependency.
(As explained in the issue, the partitioning I'm doing shouldn't have circular dependencies at all, if it weren't for the automatic inclusion of all its dependencies.)
Now when the app starts, chunk A is loading (svg is inside it and not initialized yet) -> it loads chunk B -> which has a circular dependency on A and invokes some svg functions -> crash because svg is not initialized yet.

Of course, the proper fix is that svg stays outside A and B in another chunk that doesn't depend on my app modules.

With the current way manual chunks works, this is impossible to achieve without micro-managing all modules that may have init code with `manualChunk` special cases. It doesn't really scale up.

**Comment by lukastaegert** [comment]
Yes, `manualChunks` is NOT the recommended way to create chunks! The recommended way is to just let Rollup create chunks by itself. Rollup's chunking tries to do the minimal number of chunks where it is sure it can maintain execution order.

So from this discussion I guess what you are looking for is something like "enforce that these files end up in the same chunk but do any other chunking as before for files I did not configure myself". This would require some thoughts as to what the actual algorithm is, especially with regard to execution order.

**Comment by jods4** [comment]
I can provide more context, maybe you have alternative, better, ideas.

- I'm building an application (as opposed to a library).
- It's a large app, it'll end up with 200+ pages; and it includes its own component system and more. So lots of files (last time I checked, we were at ~35 pages and had ~600 files).
- All pages are dynamically fetched with `import()` (through a router), so that the startup bundle stays small and you download only what you need, more or less.
- I'm afraid of the combinatorial explosion that would result in Rollup doing its chunking on the full graph. My understanding is that there would be like 200 "entry chunks" (kind of) and every module that is used by a different combination of dependents would create one more chunk for the combination.
- This is made worse by the fact that we can't use http/2, as we use Windows Authentication and it's not compatible. So loading tens if not hundreds of chunks doesn't sound like the best idea.
- On the other hand, the app is partitioned in several logical "areas", each targeting a specific audience. There are 20 or so area in total (so with 10 pages/each on average) and I was thinking it would made sense to bundle all pages from a single area together. This would significantly reduce the number of "entry chunks" to ~20 and would probably result in a more reasonable final graph.

One last thing is that we'd like to generate this dynamically, based on folder patterns. 
To ease development and maintenance, we don't want to a have a large central routing file where everything is listed.
This prevents us from writing "barrels" modules that re-export all the pages from each area (i.e. doing the entry chunking manually).

I suppose if we can't achieve this with rollup, we'll look into generating barrel modules for each area at compilation time.
Conceptually that's what I'm trying to do with rollup when I say I want to force some modules to be in the same chunk and have the rest of dependencies sorted out as they usually would.

> This would require some thoughts as to what the actual algorithm is, especially with regard to execution order.

I'd say: for modules that are "forced" into a shared chunk: undefined execution order (if they are independent).
It also implies they're all loaded and executed when any one module of the chunk is requested.

Other modules should be chunked exactly as they would usually be, considering the "shared" chunks as a single module that imports the union of the dependencies of every module that's inside it.

**Comment by Bulkman** [comment]
Why there is no any mention that rollup sucks at manual chunking heuristics?

**Comment by CharlieDigital** [comment]
Running into a similar issue.

The app has two sets of use cases but share the same components, model definitions, utilities, and so on.

There's some overlap in functionality so it makes sense to bundle the two use cases together.  In fact, one use case is a superset of the other (or think of it like there is a small set of features that needs to either stand alone or be part of the bigger app).

An analogy might be something like Wordpress where the blog itself is a subset of the admin.

The problem is that there's no clear way to express "preserve the default behavior *except for these entry points* which should go into a new chunk".

**Comment by unfortunately-react** [comment]
Also running into this. I would like to let Rollup decide the chunks, but put all remaining node_modules dependencies in a vendor chunk.

**Comment by mb21** [comment]
Yes, to provide another use-case for this: we have one app, with a public part and an admin-only part. I absolutely do not want to load _any_ of the JavaScript for the admin-only part into the public site, as 99% of visitors won't ever need it. It seems currently it's not possible to let rollup know about this?
